docs/csharp/roslyn-sdk/tutorials/how-to-write-csharp-analyzer-code-fix.md

```diff
@@ -0,0 +1,615 @@
+
+### Create the analyzer structure
+
+Open **MakeConstAnalyzer.cs** in Visual Studio. Change the registered action from one that acts on symbols to one that acts on syntax. In the `MakeConstAnalyzerAnalyzer.Initializer` method, find the line that registers the action on symbols:
+
+```csharp
+context.RegisterSymbolAction(AnalyzeSymbol, SymbolKind.NamedType);
+```
+
+Replace it with the following line:
+
+[!code-csharp[Register the node action](~/samples/csharp/roslyn-sdk/Tutorials/MakeConst/MakeConst/MakeConst/MakeConstAnalyzer.cs#RegisterNodeAction "Register a node action")]
+
+After that change, you can delete the `AnalyzeSymbol` method. This analyzer examines <xref:Microsoft.CodeAnalysis.CSharp.SyntaxKind.LocalDeclarationStatement?displayProperty=nameWithType>, not <xref:Microsoft.CodeAnalysis.SymbolKind.NamedType?displayProperty=nameWithType> statements. Then, put your cursor on the `AnalyzeNode` method, press **Ctrl+.** and generate the `AnalyzeNode` method.
+
+This change highlights an important design consideration for every analyzer you write: Visual Studio calls analyzers as the user is editing code. Register actions for the types of code changes that could create something your analyzer should examine. That helps minimize any performance impact for Visual Studio users that load your analyzer.
+
+Next, you should update the string resources:
+
+* Change `AnalyzerTitle` to "Variable can be made constant".
+* Change `AnalyzerMessageFormat` to "Can be made constant".
+* Change `Description` to "Make Constant".
+
+You can see the update resources in the following figure:
+
+![Update string resources](media/how-to-write-csharp-analyzer-code-fix/update-string-resources.png)
+
+Also, change the `Category` to "Usage" as shown in the following code:
+```C#
+private const string Category = "Usage";
+```
+
+When you're finished, the code in DiagnosticAnalyzer.cs should look like the following code.
+
+```C#
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace FirstAnalyzerCS
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public class MakeConstAnalyzer : DiagnosticAnalyzer
+    {
+        public const string DiagnosticId = "MakeConstCS";
+        private const string Title = "Variable can be made constant";
+        private const string MessageFormat = "Can be made constant";
+        private const string Description = "Make Constant";
```

Comment:
This code does not match the file that was generated. This is what I have after following the directions

```csharp
using System;
using System.Collections.Generic;
using System.Collections.Immutable;
using System.Linq;
using System.Threading;
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using Microsoft.CodeAnalysis.Diagnostics;

namespace MakeConst
{
    [DiagnosticAnalyzer(LanguageNames.CSharp)]
    public class MakeConstAnalyzer : DiagnosticAnalyzer
    {
        public const string DiagnosticId = "MakeConst";

        // You can change these strings in the Resources.resx file. If you do not want your analyzer to be localize-able, you can use regular strings for Title and MessageFormat.
        // See https://github.com/dotnet/roslyn/blob/master/docs/analyzers/Localizing%20Analyzers.md for more on localization
        private static readonly LocalizableString Title = new LocalizableResourceString(nameof(Resources.AnalyzerTitle), Resources.ResourceManager, typeof(Resources));
        private static readonly LocalizableString MessageFormat = new LocalizableResourceString(nameof(Resources.AnalyzerMessageFormat), Resources.ResourceManager, typeof(Resources));
        private static readonly LocalizableString Description = new LocalizableResourceString(nameof(Resources.AnalyzerDescription), Resources.ResourceManager, typeof(Resources));
        private const string Category = "Usage";

        private static DiagnosticDescriptor Rule = new DiagnosticDescriptor(DiagnosticId, Title, MessageFormat, Category, DiagnosticSeverity.Warning, isEnabledByDefault: true, description: Description);

        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get { return ImmutableArray.Create(Rule); } }

        public override void Initialize(AnalysisContext context)
        {
            // TODO: Consider registering other actions that act on syntax instead of or in addition to symbols
            // See https://github.com/dotnet/roslyn/blob/master/docs/analyzers/Analyzer%20Actions%20Semantics.md for more information
            context.RegisterSyntaxNodeAction(AnalyzeNode, SyntaxKind.LocalDeclarationStatement);
        }

        private void AnalyzeNode(SyntaxNodeAnalysisContext obj)
        {
            throw new NotImplementedException();
        }
    }
}
```